rpki-validator/src/main/java/net/ripe/rpki/validator3/api/bgp/BgpPreviewService.java

```diff
@@ -238,16 +229,82 @@ public static BgpPreviewEntry of(Asn origin, IpRange prefix, Validity validity)
             return sorting.getDirection() == Sorting.Direction.DESC ? columns.reversed() : columns;
         }
 
+        public abstract IpRange getPrefix();
+        abstract BgpPreviewEntry ofValidity(Validity validity);
+
+        public Asn getOrigin() {
+            return new Asn(Integer.toUnsignedLong(origin));
+        }
+
+    }
+
+    @lombok.Value
+    @lombok.EqualsAndHashCode(callSuper = true)
+    @lombok.ToString(callSuper = true)
+    public static class BgpPreviewEntry4 extends BgpPreviewEntry {
+        short prefixLength;
+        int prefix;
+
+        private BgpPreviewEntry4(int origin, Validity validity, short prefixLength, int prefix) {
+            super(origin, validity);
+            this.prefixLength = prefixLength;
+            this.prefix = prefix;
+        }
+
+        public static BgpPreviewEntry4 of(Asn origin, Validity validity, IpRange prefix) {
+            return new BgpPreviewEntry4((int) origin.longValue(), validity, (short) prefix.getPrefixLength(), (int) ((Ipv4Address) prefix.getStart()).longValue());
+        }
+
+        public IpRange getPrefix() {
+            return IpRange.prefix(new Ipv4Address(Integer.toUnsignedLong(prefix)), prefixLength);
+        }
+
         BgpPreviewEntry ofValidity(Validity validity) {
-            return new BgpPreviewEntry(origin, prefix, validity);
+            return new BgpPreviewEntry4(origin, validity, prefixLength, prefix);
         }
 
-        public Asn getOrigin() {
-            return new Asn(origin);
+    }
+
+    @lombok.Value
+    @lombok.EqualsAndHashCode(callSuper = true)
+    @lombok.ToString(callSuper = true)
+    public static class BgpPreviewEntry6 extends BgpPreviewEntry {
+        private static BigInteger TWO_64 = BigInteger.ONE.shiftLeft(64);
+        private static BigInteger TWO_64_MINUS_1 = TWO_64.subtract(BigInteger.ONE);
+
+        short prefixLength;
+        long prefixHi;
+        long prefixLo;
+
+        private BgpPreviewEntry6(int origin, Validity validity, short prefixLength, long prefixHi, long prefixLo) {
+            super(origin, validity);
+            this.prefixLength = prefixLength;
+            this.prefixHi = prefixHi;
+            this.prefixLo = prefixLo;
+        }
+
+        public static BgpPreviewEntry6 of(Asn origin, Validity validity, IpRange prefix) {
+            BigInteger value = prefix.getStart().getValue();
+            short prefixLength = (short) prefix.getPrefixLength();
+            long prefixHi = value.shiftRight(64).and(TWO_64_MINUS_1).longValue();
+            long prefixLo = value.and(TWO_64_MINUS_1).longValue();
```

Comment:
BigInteger.longValue: "Converts this BigInteger to a long. This conversion is analogous to a narrowing primitive conversion from long to int as defined in section 5.1.3 of The Java™ Language Specification: if this BigInteger is too big to fit in a long, only the low-order 64 bits are returned. Note that this conversion can lose information about the overall magnitude of the BigInteger value as well as return a result with the opposite sign."

Is the masking needed?

Reply:
Removed the implicit masking. I was hoping to use `longValueExact`, but that probably doesn't work when the 64-bit unsigned value doesn't fit in a 64-bit signed value (resulting in a negative value).